### PR TITLE
[quant] Add Graph Mode Passes to quantize EmbeddingBag operators

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41612 [quant] Add Graph Mode Passes to quantize EmbeddingBag operators**

Summary:
This change adds preliminary support to quantize the EmbeddingBag operators. We currently support 4-bit and 8-bit quantization+packing of the weights.

To quantize these operators, specify the operator name in the `custom_op_name` field of the NoopObserver. Based on the op name (4bit or 8bit) we call the corresponding quantization functions.
Refer to the testplan for how to invoke the qconfig for the embedding_bag ops.

Future versions of this will support 4-bit and 2-bit qtensors with native support to observe and quantize it.

NB - This version assumes that the weights in the EmbeddingBag Module reside on the same device.

Test Plan:
python test/test_quantization.py TestQuantizeDynamicJitOps.test_embedding_bag

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22609342](https://our.internmc.facebook.com/intern/diff/D22609342)

### CODE REVIEW

File: test/quantization/test_quantize_jit.py
Comment (vkuzo) [inline test/quantization/test_quantize_jit.py:2838]: optional: torch.randn to remove numpy dependency?

File: torch/csrc/jit/passes/quantization/helper.h
Comment (vkuzo) [inline torch/csrc/jit/passes/quantization/helper.h:102]: optional nit: docblock?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (vkuzo) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:252]: should this return `c10::optional`, like `findObserverName` does?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (vkuzo) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:263]: nit: brief docblock?

File: test/quantization/test_quantize_jit.py
Comment (vkuzo) [inline test/quantization/test_quantize_jit.py:2845]: makes sense that `custom_op_name` is a temporary solution.  Do we have thoughts on what to replace it with eventually /  why not now?  Would it be `EmbeddingBag{8|4|2}BitObserver` / something else?

File: torch/csrc/jit/passes/quantization/helper.h
Comment (vkuzo) [inline torch/csrc/jit/passes/quantization/helper.h:105]: what does n mean here, maybe we can name it more descriptively?

**Review by vkuzo — APPROVED** [review]
lgtm, accepting to unblock.  Feel free to wait for @jerryzh168 if a deeper review on the JIT pass is needed.

File: test/quantization/test_quantize_jit.py
Comment (jerryzh168) [reply test/quantization/test_quantize_jit.py:2845]: what's the longer term solution? are we planning to add `torch.qint4`, `torch.qint2` etc.?

File: torch/csrc/jit/passes/quantization/helper.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/helper.cpp:261]: I feel we should add some docs for these configurations to make it more clear

File: torch/csrc/jit/passes/quantization/helper.h
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/helper.h:102]: nit: I think we can remove TORCH_API actually, it's a mistake in the beginning to add TORCH_API for all these helper functions..

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:258]: follow up PR: since NoopObserver is special, probably better to add a "_" prefix to reserve this for internal use.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:268]: need to update comment

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: can we make this branch one of the inner branch of the DYNAMIC branch starting in L318?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: wait, what is this doing?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: why does this happens here?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: currently we insert prepack/unpack ops here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization/finalize.cpp#L50

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: if we need the observer info why can't we insert this op at the same place when we insert quant/dequant ops?

File: test/quantization/test_quantize_jit.py
Comment (supriyar) [reply test/quantization/test_quantize_jit.py:2845]: yes, long term this will be replaced with observer that can support `torch.qint4` and `torch.qint2`

File: torch/csrc/jit/passes/quantization/helper.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/helper.cpp:261]: Do you mean adding the op signature here for every op?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: That's mainly because the logic where we insert quant/dequant ops doesn't get called for every module instance. I added it here to make sure the prepack and quant ops are added for each embeddingBag module instance.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: `insertEmbeddingBagQuantOps` below is run only for the weight tensors. For the indices input we still have an observer attached which we need to disregard. This part of code replaces that observer output with the original value so that the observer can be deleted in cleanup step. I'll update comment to make it more descriptive.

File: torch/csrc/jit/passes/quantization/helper.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/helper.cpp:261]: yeah

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: why can't we skip inserting observer for input? I remember we implemented that support before for batchnorm?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#torch::jit::_observe_inputs_call_func

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: the current logic is we insert for each graph, but the graph itself might be shared by multiple module instances, why that does not work?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: I think we spoke offline about this as well but this case is special so we need to observe both. If we don't I run into 
```
RuntimeError: branches for `if` should return values that are observed consistently, if node:%input.22 : Tensor, %weight : Tensor = prim::If(%19) # /home/supriyar/local/miniconda3/envs/pytorch/lib/python3.7/site-packages/torch/nn/functional.py:1908:4
```

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: My understanding was that we cannot have a shared graph here as each module instance will need its own set of pack and quant ops inserted. I initially did take this approach but I noticed that in the final graph only one of the embedding modules had the quant ops inserted.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:310]: Oh I see, then could you add a comment for this?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: you mean only one of the embedding modules has the packing? or only one has quant + packing and the other one is just ignored?

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: I don't see why we can't have a shared graph here, could you show the problem with putting this together with insert quant/dequant in more detail?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1319]: I think it's better to keep the insert here. breaking this insert means there is something wrong and we'll need to fix that

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1319]: when is the qparam map empty?

Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1319]: Noop observer is an example of an observer that doesn't have qparam_map

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1319]: OK, it's probably fine in this case, could you also add a comment for this check?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:863]: could you also add a `WithInsertPoint`? we want to insert before the use node I think.

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:863]: should just be: `WithInsertPoint ins(embedding_bag_loat_op);`

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:860]: what are the possible cases here? can we have stronger guarantees here? is the observer_out always going to be used by an embedding_bag op here?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:1296]: @jerryzh168 I think it should be fine. I realized my test graph wasn't doing the right thing. Will update to move to insert quant_dequant function.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:863]: We've already added the insert point in insertQuantizationOps at the output of the observer node.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:384]: can you merge this branch with the one lin L399 now?

Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:384]: I prefer keeping it separate since this is a special case and this code will be removed in the future. Wanted to make that a little more obvious :)

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:390]: these two lines are the same as 417-419

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:384]: might also be good to have a isEmbeddingBagOp function to be consistent with isFP16NoopObserver

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:409]: btw do we need the `quant` value here?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:271]: suggest to rename this to insertQuantDeQuant

**Review by jerryzh168 — APPROVED** [review]
Looks good! thanks. had a few more inline comments.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:222]: btw we can check for types now I think

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:384]: I see, sure. do you mean we plan to remove the swapping of input and weight in the future in the embedding bag module? I think this will be needed in graph mode if that does not change

**Pull request closed without merging**
